### ISSUE TITLE
DsiWin32 AffinityMask functions should use DWORD_PTR

### ISSUE THREAD

**Issue by PatrickvL** [gabr42/GpDelphiUnits#6]
From : https://code.google.com/p/omnithreadlibrary/issues/detail?id=79

What steps will reproduce the problem?
1. On systems with more than 32 logical processors, OmniThread library doesn't start because of range check errors in AffinityMask functions
2.
3.

What is the expected output? What do you see instead?

DwiWin32's wrappers for AffinityMask functions should be upgraded to the actual Win32 API (using DWORD_PTR instead of DWORD).
Also, DSiCPUIDs should be extended to support 64 processors.
Also, loops reading DSiCPUIDs should use High(DSiCPUIDs) instead of 32 (or 64)

What version of the product are you using? On what operating system?
Most recent code. Host with 32 processors.

Please provide any additional information below.
Note, this happens on 64 bit using XE2, but probably also applies to all other supported compiler versions, and might even fail on 32 bit Windows OSes (although it's highly unlikely those support more than 32 processors).

**Comment by gabr42** [comment]
You're correct, this indeed has to be fixed. I don't notice such problems as I don't yet have a computer with > 32 logical processors.

**Comment by gabr42** [comment]
DSiWin32 already uses Win32 API call with DWORD_PTR parameters. It's just the mappers that map this to/from string that can't handle more than 32 cores.

**Comment by gabr42** [comment]
Fixed. Please check if it works for you.

**Comment by PatrickvL** [comment]
We finally got around to testing your fix, and there's good news and bad news to report:

The good news is, we got it to work, your 64 bit support works as excepted.

But the bad news is, we had to fix a Delphi RTL bug to get it to work!

The thing is, DSiGetThreadAffinityMask must be implemented using kernel32's SetThreadAffinityMask, since there's no GetThreadAffinityMask. Now, the problem we encountered is, that Delphi's RTL (at least, versions up to XE2) had a bug in the signature used for importing SetThreadAffinityMask; MSDN says it requires a DWORD_PRT, but the Delphi RTL specifies a DWORD. If you call this on a system with more than 32 cores (the one we tested on had 36 cores) this crashes in such a way that the entire process disappears.

Once we spotted the error and used our own import using the correct DWORD_PTR type, all went fine.

Hopefully this report helps.

Cheers,
Patrick van Logchem
Every Angle BV Nederland

**Comment by PatrickvL** [comment]
For more information, see : http://qc.embarcadero.com/wc/qcmain.aspx?d=106659 and http://qc.embarcadero.com/wc/qcmain.aspx?d=121460

**Comment by gabr42** [comment]
Thanks for the report, I'll incorporate a similar workaround into DSiWin32.
